website/project/views/comment.py

```diff
@@ -198,41 +307,85 @@ def is_reply(target):
 def list_comments(auth, **kwargs):
     node = kwargs['node'] or kwargs['project']
     anonymous = has_anonymous_link(node, auth)
+    page = request.args.get('page')
     guid = request.args.get('target')
-    target = resolve_target(node, guid)
-    serialized_comments = serialize_comments(target, auth, anonymous)
-    n_unread = 0
+    root_id = request.args.get('rootId')
+    is_list = request.args.get('isCommentList')
+    is_widget = False
+
+    if page == 'total' and not root_id:  # "Total" on discussion page
+        comments = list_total_comments(node, auth, 'total')
+    elif page == 'total':  # Discussion widget on overview's page
+        comments = list_total_comments_widget(node, auth)
+        is_widget = True
+    elif not root_id:  # Overview/Files/Wiki page on discussion page
+        comments = list_total_comments(node, auth, page)
+    else:
+        target = resolve_target(node, page, guid)
+        comments = getattr(target, 'commented', [])
+
+    if is_list:
+        discussions = comment_discussion(comments, node, anonymous=anonymous, widget=is_widget)
 
+    n_unread = 0
     if node.is_contributor(auth.user):
         if auth.user.comments_viewed_timestamp is None:
             auth.user.comments_viewed_timestamp = {}
             auth.user.save()
-        n_unread = n_unread_comments(target, auth.user)
-    return {
-        'comments': serialized_comments,
+        n_unread = n_unread_comments(node, auth.user, page, root_id)
+
+    ret = {
+        'comments': [
+            serialize_comment(comment, auth, anonymous)
+            for comment in comments
+        ],
         'nUnread': n_unread
     }
+    if is_list:
+        ret.update(discussions)
+    return ret
+
+
+def list_total_comments_widget(node, auth):
+    comments_keys = Comment.find(Q('node', 'eq', node)).get_keys()
+    comments = []
+    for cid in comments_keys:
+        cmt = Comment.load(cid)
```

Comment:
Why use `get_keys` if you are going to load every `Comment` anyway? Could you do:

```python
comments = list(Comment.find(Q('node', 'eq', node)))
```